src/optimizer/pushdown/pushdown_cross_product.cpp

```diff
@@ -37,16 +43,36 @@ unique_ptr<LogicalOperator> FilterPushdown::PushdownCrossProduct(unique_ptr<Logi
 				if (join_ref_type == JoinRefType::ASOF) {
 					// AsOf is really a table lookup, so we don't push filters
 					// down into the lookup (right) table
-					join_expressions.push_back(std::move(f->filter));
+					asof_filters.push_back(std::move(f));
```

Comment:
For other joins we need to distinguish between filters in the `ON` clause (join conditions) and filters in the `WHERE` clause. The `ON` clause needs to be applied to the matching process, whereas the `WHERE` clause needs to be applied after the join. For inner joins this does not matter as these are equivalent, but it matters for other types of joins. For example, for left joins:

```sql
D FROM (SELECT 42) t(a) LEFT JOIN (SELECT 84) t2(b) ON (a>b AND b<50);
┌───────┬───────┐
│   a   │   b   │
│ int32 │ int32 │
├───────┼───────┤
│    42 │  NULL │
└───────┴───────┘
D FROM (SELECT 42) t(a) LEFT JOIN (SELECT 84) t2(b) ON (a>b) WHERE (b<50);
┌───────┬───────┐
│   a   │   b   │
│ int32 │ int32 │
├───────┴───────┤
│    0 rows     │
└───────────────┘
```

It seems that currently these are treated as the same in the `ASOF` join. This causes some odd behavior, for example, the ASOF join in the example returns this result:

```sql
SELECT 
      a.seq_no,
      a.amount,
      b.amount
  FROM issue13899 AS a 
  ASOF JOIN issue13899 AS b
      ON a.seq_no>=b.seq_no
    ;
┌────────┬───────────────┬───────────────┐
│ seq_no │    amount     │    amount     │
│ int32  │ decimal(10,2) │ decimal(10,2) │
├────────┼───────────────┼───────────────┤
│      1 │          1.00 │          1.00 │
│      2 │          NULL │          NULL │
│      3 │          NULL │          NULL │
│      4 │          NULL │          NULL │
│      5 │          2.00 │          2.00 │
│      6 │          NULL │          NULL │
│      7 │          NULL │          NULL │
│      8 │          3.00 │          3.00 │
│      9 │          NULL │          NULL │
│     10 │          NULL │          NULL │
│     11 │          5.00 │          5.00 │
├────────┴───────────────┴───────────────┤
│ 11 rows                      3 columns │
└────────────────────────────────────────┘
```

Now, if I add the filter `b.amount IS NOT NULL`, I would expect to remove all of the rows with `NULL`. Instead, the `WHERE` clause appears to be pushed down into the filter and the matches *themselves* are changed.

```sql
SELECT 
        a.seq_no,
        a.amount,
        b.amount
    FROM issue13899 AS a 
    ASOF JOIN issue13899 AS b
        ON a.seq_no>=b.seq_no
    WHERE b.amount IS NOT NULL;
┌────────┬───────────────┬───────────────┐
│ seq_no │    amount     │    amount     │
│ int32  │ decimal(10,2) │ decimal(10,2) │
├────────┼───────────────┼───────────────┤
│      1 │          1.00 │          1.00 │
│      2 │          NULL │          1.00 │
│      3 │          NULL │          1.00 │
│      4 │          NULL │          1.00 │
│      5 │          2.00 │          2.00 │
│      6 │          NULL │          2.00 │
│      7 │          NULL │          2.00 │
│      8 │          3.00 │          3.00 │
│      9 │          NULL │          3.00 │
│     10 │          NULL │          3.00 │
│     11 │          5.00 │          5.00 │
├────────┴───────────────┴───────────────┤
│ 11 rows                      3 columns │
└────────────────────────────────────────┘
```

I think the issue here lies in the planning/binding phase rather than in the optimizer phase - we need to somehow preserve where filters come from (either the join conditions, or the `WHERE` clause). I also see that AsOf joins are only created from a `LogicalComparisonJoin` - perhaps they should instead be converted to support the `LogicalAnyJoin`? We can then do the handling of arbitrary expressions in the `ON` clause during the planning phase (by e.g. pushing them into the right side as is being done here now in an optimizer).

Reply:
I think the correct approach is to always treat AsOf like a kind of left join.